runtime/sema/check_function.go

```diff
@@ -173,11 +182,17 @@ func (checker *Checker) checkFunction(
 			functionActivation.InitializationInfo = initializationInfo
 
 			if functionBlock != nil {
-				checker.visitFunctionBlock(
-					functionBlock,
-					functionType.ReturnTypeAnnotation,
-					checkResourceLoss,
-				)
+				isPure := checker.InNewPurityScope(functionType.Purity == PureFunction, func() {
+					checker.visitFunctionBlock(
+						functionBlock,
+						functionType.ReturnTypeAnnotation,
+						checkResourceLoss,
+					)
+				})
+
+				if functionType.Purity == UnknownPurity {
```

Comment:
This is the core of the analysis; we check the function in a new "purity scope", which starts pure and then will be monotonically updated to "impure" as soon as an impurity is observed. If we are enforcing purity, as we would do if the function has an explicit `pure` annotation, then this will cause a type error. Otherwise if the function has no annotation we update its purity to be the purity of the body block.